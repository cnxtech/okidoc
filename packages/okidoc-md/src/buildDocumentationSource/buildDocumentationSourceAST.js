import fs from 'fs';
import path from 'path';
import * as t from '@babel/types';
import traverseSource from '../utils/traverseSource';
import traverseEntries from '../utils/traverseEntries';
import traverseFiles from '../utils/traverseFiles';
import {
  createApiVisitor,
  createApiMethod,
  createApiInterface,
  buildApiClassDeclaration,
} from '../api';

function ensureSourceOptionsValid({ entry, pattern, source }) {
  if (!entry && !pattern && !source) {
    throw new Error(
      `'entry' or/and 'pattern' or 'source' options should be defined`,
    );
  }

  if (source && (entry || pattern)) {
    throw new Error(
      `if 'source' options provided, 'entry' and 'pattern' should be undefined`,
    );
  }
}

function ensureVisitorExists(apiVisitorPath) {
  if (!fs.existsSync(apiVisitorPath)) {
    throw new Error(
      `'visitor' option should be valid js file path. '${apiVisitorPath}' is not exists`,
    );
  }
}

function ensureCreateApiVisitorIsFunction(createCustomApiVisitor) {
  if (typeof createCustomApiVisitor !== 'function') {
    throw new Error(
      `'visitor' file should export function 'createApiVisitor' (exports.createApiVisitor = ...)`,
    );
  }
}

function ensureCustomApiVisitorIsCorrect(
  visitors,
  customApiVisitor,
  visitorPath,
) {
  const visitorsKeys = Object.keys(visitors);
  const customApiVisitorKeys = Object.keys(customApiVisitor);
  const intersection = customApiVisitorKeys.filter(key =>
    visitorsKeys.includes(key),
  );

  if (intersection.length) {
    console.error(
      `default config visitors ${JSON.stringify(
        intersection,
      )} are overridden by config from 'visitor.createApiVisitor' (${visitorPath})`,
    );
  }
}

function traverse({ entry, pattern, source }, visitors) {
  if (source) {
    traverseSource(source, visitors);
    return;
  }

  if (entry) {
    traverseEntries(entry, visitors, { pattern });
    return;
  }

  traverseFiles(pattern, visitors);
}

function buildDocumentationSourceAST({
  entry,
  pattern,
  source,
  tag,
  visitor: visitorPath,
}) {
  const apiInterfaces = [];
  const classApiMethods = [];
  const functions = [];

  ensureSourceOptionsValid({ entry, pattern, source });

  if (!tag && !visitorPath) {
    throw new Error(
      `'tag' or/and 'visitor' options should be defined ('${pattern}')`,
    );
  }

  // NOTE: read about visitors
  // https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-visitors
  // https://github.com/babel/babel/blob/master/packages/babel-traverse/src/visitors.js
  const visitors = {
    'InterfaceDeclaration|TSInterfaceDeclaration': path => {
      apiInterfaces.push(createApiInterface(path.node));
    },
  };

  if (tag) {
    Object.assign(
      visitors,
      createApiVisitor(tag, path => {
        if (t.isClassMethod(path.node || t.isClassProperty(path.node))) {
          classApiMethods.push(createApiMethod(path.node, tag));
        }

        if (
          t.isVariableDeclaration(path.node) ||
          t.isExportNamedDeclaration(path.node) ||
          t.isFunctionDeclaration(path.node)
        ) {
          functions.push(createApiMethod(path.node, tag));
        }
      }),
    );
  }

  if (visitorPath) {
    visitorPath = path.resolve(visitorPath);

    ensureVisitorExists(visitorPath);

    const {
      createApiVisitor: createCustomApiVisitor,
      createApiMethod: createCustomApiMethod,
    } = require(visitorPath);

    ensureCreateApiVisitorIsFunction(createCustomApiVisitor);

    const customApiVisitor = createCustomApiVisitor(path => {
      if (t.isClassMethod(path.node || t.isClassProperty(path.node))) {
        classApiMethods.push(
          typeof createCustomApiMethod === 'function'
            ? createCustomApiMethod(path.node)
            : createApiMethod(path.node),
        );
      }
    });

    ensureCustomApiVisitorIsCorrect(visitors, customApiVisitor, visitorPath);

    Object.assign(visitors, customApiVisitor);
  }

  traverse({ entry, pattern, source }, visitors);

  const apiClassDeclaration = buildApiClassDeclaration(classApiMethods);

  return {
    type: 'Program',
    body: apiInterfaces.concat(apiClassDeclaration, functions),
  };
}

export default buildDocumentationSourceAST;
