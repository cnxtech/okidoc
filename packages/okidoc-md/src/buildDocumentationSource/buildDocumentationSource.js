import generate from '@babel/generator';

import buildDocumentationSourceAST from './buildDocumentationSourceAST';

function buildDocumentationSource({
  entry,
  source,
  pattern,
  tag,
  visitor,
  interfaces = {},
}) {
  return generate(
    buildDocumentationSourceAST({
      entry,
      source,
      pattern,
      tag,
      visitor,
      interfaces,
    }),
  ).code;
}

export default buildDocumentationSource;
