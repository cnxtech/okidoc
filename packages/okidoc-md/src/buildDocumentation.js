import buildDocumentationSource from './buildDocumentationSource';
import buildDocumentationMarkdown from './utils/buildDocumentationMarkdown';

function buildDocumentation({
  title,
  entry,
  source,
  pattern,
  tag,
  visitor,
  interfaces,
  excludeKind = [],
}) {
  const documentationSource = buildDocumentationSource({
    entry,
    source,
    pattern,
    tag,
    visitor,
    interfaces,
  });

  return buildDocumentationMarkdown(documentationSource, {
    title,
    sourceInfo: entry || pattern,
    excludeKind,
  });
}

export default buildDocumentation;
