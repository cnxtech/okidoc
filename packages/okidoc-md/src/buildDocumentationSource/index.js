import buildDocumentationSource from './buildDocumentationSource';

export default buildDocumentationSource;
