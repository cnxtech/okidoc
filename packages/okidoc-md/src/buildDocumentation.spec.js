import buildDocumentation from './buildDocumentation';
import buildDocumentationSource from './buildDocumentationSource';

describe('buildDocumentation', () => {
  it('should build documentation for class ', async () => {
    const sourceCode = `
      /**
      * @doc UI
      */
      class C1 {
        m1() {}
      }
      
      /**
      * @doc UI
      */
      export class C2 {
        m2() {}
      }
      
      /**
      * @doc UI
      */
      export default class C3 {
        m3() {}
      }
    `;

    expect(
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      }),
    ).toMatchSnapshot();
  });

  it('should build documentation for functions ', async () => {
    const sourceCode = `
      /**
      * @doc UI
      */
      function functionDeclaration() {}
      
      /**
      * @doc UI
      */
      var varFunctionExpression = function () {}
      
      /**
      * @doc UI
      */
      let letFunctionExpression = function () {}

      /**
      * @doc UI
      */
      const constFunctionExpression = function () {}
      
      /**
      * @doc UI
      */
      const constArrowFunction = () => {}
      
      /**
      * @doc UI
      */
      export const exportConstArrowFunction = () => {}
      
      /**
      * @doc UI
      */
      export default exportDefaultConstArrowFunction = () => {}
    `;

    expect(
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      }),
    ).toMatchSnapshot();
  });

  it('should show readable error from `documentation.js`', async () => {
    const sourceCode = `
      /**
      * @doc UI
      */
      class UI {
        public show() {}
      }
    `;

    expect.assertions(1);

    try {
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});
