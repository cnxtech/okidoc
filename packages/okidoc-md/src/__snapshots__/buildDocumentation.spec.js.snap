// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildDocumentation should build documentation for class  1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Documentation

## m1()

## m2()

## m3()
"
`;

exports[`buildDocumentation should build documentation for functions  1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Documentation

## functionDeclaration()

## varFunctionExpression()

## letFunctionExpression()

## constFunctionExpression()

## constArrowFunction()

## exportConstArrowFunction()

## exportDefaultConstArrowFunction()
"
`;

exports[`buildDocumentation should show readable error from \`documentation.js\` 1`] = `
[SyntaxError: documentation source: Unexpected token, expected , (6:19)
  4 | /**
  5 | */
> 6 | function fooGood<T extends {
    |                   ^
  7 |   x: number;
  8 | }>(obj: T): T {}]
`;
